app/SocialMedia/Twitter.php

```diff
@@ -0,0 +1,255 @@
+	 * @throws \App\Exceptions\AppException
+	 */
+	public function __construct(string $userName)
+	{
+		$this->userName = $userName;
+		if (!\is_object(static::$twitterConnection)) {
+			$configTitter = \App\SocialMedia::getInstance(static::$socialMediaType);
+			static::$twitterConnection = new \Abraham\TwitterOAuth\TwitterOAuth(
+				$configTitter->get('twitter_api_key'),
+				$configTitter->get('twitter_api_secret')
+			);
+			static::$twitterConnection->setDecodeJsonAsArray(true);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function retrieveDataFromApi()
+	{
+		if (!\App\RequestUtil::isNetConnection()) {
+			$this->logErrorDb('No internet connection');
+			return;
+		}
+		$db = \App\Db::getInstance();
+		$maxId = (new \App\Db\Query())
+			->from('u_#__social_media_twitter')
+			->where(['twitter_login' => $this->userName])
+			->max((new \yii\db\Expression('CAST(id_twitter AS INT)')));
+		$param['screen_name'] = $this->userName;
+		$indexOfText = 'text';
+		if (static::TWEET_MODE === 'extended') {
+			$param['tweet_mode'] = 'extended';
+			$indexOfText = 'full_text';
+		}
+		if (!\is_null($maxId)) {
+			//Retrieve only new data from Api
+			//"There are limits to the number of Tweets that can be accessed through the API. If the limit of Tweets has occured since the since_id, the since_id will be forced to the oldest ID available."
+			$param['since_id'] = $maxId;
+		}
+		$this->logInfoDb('Begin downloading new messages');
+		$allMessages = $this->getFromApi('statuses/user_timeline', $param);
+		$cnt = 0;
+		foreach ($allMessages as $rowTwitter) {
+			$rowTwitter['id'] = \App\Purifier::encodeHtml($rowTwitter['id']);
+			$rowTwitter['created_at'] = \App\Purifier::encodeHtml($rowTwitter['created_at']);
+			$rowTwitter[$rowTwitter[$indexOfText]] = \App\Purifier::encodeHtml($rowTwitter[$indexOfText]);
+			if (!isset($rowTwitter['user']['name'])) {
+				throw new \App\Exceptions\AppException('ERR_NO_VALUE||"user name"');
+			}
+			$rowTwitter['user']['name'] = \App\Purifier::encodeHtml($rowTwitter['user']['name']);
+			if (!(new \App\Db\Query())->from('u_#__social_media_twitter')->where(['id_twitter' => $rowTwitter['id']])->exists()) {
+				$db->createCommand()->insert('u_#__social_media_twitter', [
+					'id_twitter' => $rowTwitter['id'],
+					'twitter_login' => $this->userName,
+					'twitter_name' => $rowTwitter['user']['name'],
+					'message' => $rowTwitter[$indexOfText],
+					'created' => \DateTimeField::convertToUserTimeZone($rowTwitter['created_at'])->format('Y-m-d H:i:s'),
+				])->execute();
+				$cnt++;
+			}
+		}
+		if ($cnt > 0) {
+			$this->logInfoDb($cnt . ' new messages downloaded from');
+		} else {
+			$this->logInfoDb('No new messages');
+		}
+	}
+
+	/**
+	 * Log info.
+	 *
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	private function logInfoDb(string $message)
+	{
+		if (static::$logOn) {
+			static::logInfo('[' . $this->userName . ']: ' . $message);
```

Comment:
('[' . $this->userName . ']: ' . $message) = > ("[{$this->userName}]: {$message})
and other place too